AS-709 Handle exceptions from WSM

ungwudik opened the pull request:
- Handle exceptions from WSM endpoints
- Also, return 404 (and handle exceptions) instead of 500 when attempting to update a non-existent snapshot

Created AS-718 as follow-up work.

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

ungwudik committed:
AS-709 Handle exceptions from WSM
Return 404 when attempting to update a non-existent snapshot

ungwudik commented:
Jenkins Automation Tests failed, but those don't look related to this PR. I assume it's due to flakiness.

sagehen03 commented on core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotService.scala:
```diff
@@ -107,6 +107,7 @@ class SnapshotService(protected val userInfo: UserInfo, val dataSource: SlickDat
   def updateSnapshot(workspaceName: WorkspaceName, snapshotId: String, updateInfo: UpdateDataReferenceRequestBody): Future[Unit] = {
     val snapshotUuid = validateSnapshotId(snapshotId)
     getWorkspaceContextAndPermissions(workspaceName, SamWorkspaceActions.write, Some(WorkspaceAttributeSpecs(all = false))).map { workspaceContext =>
+      workspaceManagerDAO.getDataReference(workspaceContext.workspaceIdAsUUID, snapshotUuid, userInfo.accessToken)
```
This seems unusual to me and unrelated to handling WSM exceptions.  Is there some necessary side effect we get from calling `getDataReference`?

davidangb replied:
`updateDataReference` returns a 20x even if the reference does not exist, so we need `getDataReference` here in order to generate the 404 in the case of nonexistent references.

these WSM APIs are in the process of being deprecated and replaced by newer-version APIs, so I don't think it's worth changing the WSM behavior. That said, we should probably add explanatory code comments/TODOs here.

sagehen03 commented on core/src/main/scala/org/broadinstitute/dsde/rawls/webservice/RawlsApiService.scala:
```diff
@@ -43,6 +44,8 @@ object RawlsApiService extends LazyLogging {
       case rollback:MySQLTransactionRollbackException =>
         logger.error(s"ROLLBACK EXCEPTION, PROBABLE DEADLOCK: ${rollback.getMessage} [${rollback.getErrorCode} ${rollback.getSQLState}] ${rollback.getNextException}", rollback)
         complete(StatusCodes.InternalServerError -> ErrorReport(rollback))
+      case wsmApiException: ApiException =>
```
seems reasonable to me :)

sagehen03 reviewed (commented):
I'm still new to Scala and Rawls, but I think it would be nice to add a unit test to confirm that you're getting the change you expect.  At the top of `SnapshotApiServiceSpec` you should be able to  override a method like getDataReference and have that method for a particular argument throw a ApiException.  Let me know if that sounds like something you'd be interested in and I can try to add more detail.

davidangb reviewed (commented):
code change looks good. Are there any tests that can/should be written or updated?

ungwudik committed:
Update mock for updateDataReference function
Add TODOs

ungwudik commented:
> I'm still new to Scala and Rawls, but I think it would be nice to add a unit test to confirm that you're getting the change you expect. At the top of `SnapshotApiServiceSpec` you should be able to override a method like getDataReference and have that method for a particular argument throw a ApiException. Let me know if that sounds like something you'd be interested in and I can try to add more detail.

We do have a mocked [unit test here](https://github.com/broadinstitute/rawls/blob/09d8572256be1a06e0bc44a95e0220e325f46d7d/core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/SnapshotApiServiceSpec.scala#L339) (see P.S./context below) for verifying that attempting to update a non-existent snapshot returns a `404`. But if you meant specifically testing that `bio.terra.workspace.client.ApiException` is thrown, then we can also discuss that/its value.


**P.S.**
Some context that may help: The unit test in `SnapshotApiServiceSpec` was (stay with me) ahead of the `updateSnapshot` function that it's testing. This is because:
1. The function was originally returning 404 as expected, and this behavior matched the test. So far, so good.
2. Then, the behavior changed in WSM causing it to start returning 204 when updating a non-existent snapshot. This 204 bubbled up to Rawls as well.
3. But after #2 happened, the mocked test stayed the same (i.e. still returning the expected 404). This meant there was a mismatch between the functionality AND what the test was doing.
4. We later decided that Rawls should return 404 even if WSM returns a 204. This is why I'm updating the functionality here, but the test already matches the expectation.

Let me know if you have any questions or need clarification regarding this context. I'm happy to discuss offline as well.

ungwudik commented:
> code change looks good. Are there any tests that can/should be written or updated?

See my response to Drew about test [here](https://github.com/broadinstitute/rawls/pull/1410#issuecomment-815973602). That said, I've updated `MockWorkspaceManagerDAO.scala` to remove the now-unnecessary `throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.NotFound, "Not found"))`, and have also added a condition for updating snapshot reference.

ungwudik commented:
> > code change looks good. Are there any tests that can/should be written or updated?
> 
> See my response to Drew about test [here](https://github.com/broadinstitute/rawls/pull/1410#issuecomment-815973602). That said, I've updated `MockWorkspaceManagerDAO.scala` to remove the now-unnecessary `throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.NotFound, "Not found"))`, and have also added a condition for updating snapshot reference.

Let me know if you have further feedback.

sagehen03 commented:
That all makes sense @ungwudik  Thanks for the explanation approving now!  I do think it would make sense to remove the RawlsException being thrown from the mock since that should happen via the Rawls exception handler now.

davidangb reviewed (approved):
:+1: Updating the mock to not throw 404 is great, thank you!
